### DISCUSSION TITLE
SQLAlchemy Core - JOIN support with USING keyword

### DISCUSSION THREAD

**Discussion by njiang747** [sqlalchemy/sqlalchemy#6186] [Usage Questions] [answered]
# Overview

I would like to perform joins with `USING` in SQLAlchemy Core. My particular use case is for BigQuery but this keyword also exists in other SQL languages like PostgreSQL, MySQL, and Oracle SQL.

The existing [join method](https://docs.sqlalchemy.org/en/14/core/selectable.html#sqlalchemy.sql.expression.FromClause.join) only supports an `onclause`.

# Desired Solution

Ideally I'd like it to support another alternative parameter `using_column_names: Iterable[str]` that specifies the column names to join over in the `USING` clause.

Ex:
```python
table_1.join(table_2, using_column_names=['col_1', 'col_2'])
```
should result in:
```sql
table_1 JOIN table_2 USING (col_1, col_2)
```

# Motivating Example

### Table 1 (t1)
| c1 | c2 | x  |
| -- | -- | -- |
| 0  | 0  | 0  |
| 0  | 1  | 1  |
| 1  | 0  | 2  |

### Table 2 (t2)
| c1 | c2 | y  |
| -- | -- | -- |
| 0  | 0  | 3  |
| 1  | 1  | 4  |
| 1  | 2  | 5  |

### Table 3 (t3)
| c1 | c2 | z  |
| -- | -- | -- |
| 0  | 0  | 6  |
| 1  | 1  | 7  |
| 2  | 2  | 8  |

## Desired Query

```sql
SELECT
  c1, c2, x, y, z
FROM
  t1
FULL OUTER JOIN
  t2
USING (c1, c2)
FULL OUTER JOIN
  t3
USING (c1, c2)
```

which yields (`-` indicates `null`):

| c1 | c2 | x  | y  | z  |
| -- | -- | -- | -- | -- |
| 0  | 0  | 0  | 3  | 6  |
| 0  | 1  | 1  | -  | -  |
| 1  | 0  | 2  | -  | -  |
| 1  | 1  | -  | 4  | 7  |
| 1  | 2  | -  | 5  | -  |
| 2  | 2  | -  | -  | 8  |

## Current Workaround

The workaround that I've found so far that works is:

```sql
SELECT
  COALESCE(t1.c1, t2.c1, t3.c1) AS c1,
  COALESCE(t1.c2, t2.c2, t3.c2) AS c1,
  x, y, z
FROM
  t1
FULL OUTER JOIN
  t2
ON
  t1.c1 = t2.c2
  AND t1.c2 = t2.c2
FULL OUTER JOIN
  t3
ON
  COALESCE(t1.c1, t2.c1) = t3.c1
  AND COALESCE(t1.c2, t2.c2) = t3.c2
```

The problem is that this query is super verbose and especially so if we want to do a ton of consecutive joins like this.

**Comment by zzzeek** [comment, marked as answer]
took a look at USING and it appears to be based on naming conventions, which makes this a poor candidate for a SQLAlchemy feature since it would be brittle and break as soon as SQLAlchemy applied any kind of labeling to columns in subqueries.

However, the USING(x, y) syntax follows normal [function parenthesized form](https://docs.sqlalchemy.org/en/14/tutorial/data.html#working-with-sql-functions), so without even having to use a [custom SQL construct](https://docs.sqlalchemy.org/en/14/core/compiler.html) your SQL can be rendered as follows :

```python
from sqlalchemy import column
from sqlalchemy import func
from sqlalchemy import select
from sqlalchemy import table


table_1 = table("t1", column("c1"), column("c2"), column("x"))
table_2 = table("t2", column("c1"), column("c2"), column("y"))
table_3 = table("t3", column("c1"), column("c2"), column("z"))


c1, c2 = column("c1"), column("c2")
stmt = (
    select(c1, c2, table_1.c.x, table_2.c.y, table_3.c.z)
    .join_from(
        table_1, table_2, func.using(c1, c2), full=True
    )
    .join(table_3, func.using(c1, c2), full=True)
)

print(stmt)
```

output:

```sql
SELECT c1, c2, t1.x, t2.y, t3.z 
FROM t1 FULL OUTER JOIN t2 ON using(c1, c2) FULL OUTER JOIN t3 ON using(c1, c2)
```

**Reply by njiang747** [reply]
@zzzeek Just realized this is not correct syntax. `USING (c1, c2)` should be a substitute for `ON`, `ON USING` is invalid.

**Reply by zzzeek** [reply]
OK then lets do custom compilation:


```python
from sqlalchemy import column
from sqlalchemy import func
from sqlalchemy import select
from sqlalchemy import table
from sqlalchemy.ext.compiler import compiles
from sqlalchemy.sql import Join
from sqlalchemy.sql.functions import FunctionElement


@compiles(Join)
def _setup_using(element, compiler, **kw):
    text = compiler.visit_join(element, **kw)
    if (
        isinstance(element.onclause, FunctionElement)
        and element.onclause.name == "using"
    ):
        text = text.replace("ON using", "USING")

    return text


table_1 = table("t1", column("c1"), column("c2"), column("x"))
table_2 = table("t2", column("c1"), column("c2"), column("y"))
table_3 = table("t3", column("c1"), column("c2"), column("z"))


c1, c2 = column("c1"), column("c2")
stmt = (
    select(c1, c2, table_1.c.x, table_2.c.y, table_3.c.z)
    .join_from(table_1, table_2, func.using(c1, c2), full=True)
    .join(table_3, func.using(c1, c2), full=True)
)

print(stmt)
```

that swaps out "ON using" for "USING".

the reason adding this as a kw to Join is problematic is because that doesn't provide a way of getting at these columns for the columns clause, and additionally we are moving away from table.join() as the usual API for joins in favor of select.join_from().    "USING" is a very seldom used type of construct so it makes sense this would at most be a special Using() kind of object that would be interpreted similarly as above.

**Reply by njiang747** [reply]
This solves my problem. Thanks!

The concerns about `using` as a keyword make sense, thanks for the explanation.

**Comment by njiang747** [comment]
@zzzeek Thanks for the response! I will test this out!

**Comment by sebastian-correa** [comment]
I was trying to do the same as @njiang747 and @zzzeek's solution didn't work as-is for me. Using `select(...).join_from(...)` left the `ON using` in my case, but doing `select(...).select_from(table_1.join(table_2, ....))` worked. Not sure if it's expected, but this might save time to other people looking for this solution.

Other than this, it seems that the best course of action is something like this code that I did not test:

```python
outerjoin_tables = [Model1, Model2, Model3, Model4]

query = select(func.coalesce(*[model.join_col for model in outerjoin_tables]), ...)

for i, model in enumerate(outerjoin_tables):
    query = query.join(model, onclause=func.coalesce(*[model.join_col for model in outerjoin_tables[:i + 1]]) == model.join_col, full=True)
```
